pkg/storage/stores/shipper/bloomshipper/shipper.go

```diff
@@ -59,14 +58,8 @@ func (s *Shipper) Fetch(ctx context.Context, tenantID string, blocks []BlockRef,
 		select {
 		case <-ctx.Done():
 			return fmt.Errorf("failed to fetch blocks: %w", ctx.Err())
-		case result, ok := <-blocksChannel:
-			if !ok {
-				return nil
-			}
-			err := runCallback(callback, result)
-			if err != nil {
-				return err
-			}
+		case result := <-blocksChannel:
+			return runCallback(callback, result)
```

Comment:
This will only work for downloading a single block. I want to fix this in a follow up PR.

Reply:
Good point. I think the actuall problem was only
```
case result, ok := <-blocksChannel:
```
which should be

```
case result := <-blocksChannel:
```

Reply:
> what was the problem with the previous implementation?

The previous implementation never exited from the `for { select {} }` loop, because `runCallback()` should not return an error.